### DISCUSSION TITLE
4 Table relationship using secondary

### DISCUSSION THREAD

**Discussion by desktilldawn** [sqlalchemy/sqlalchemy#11591] [Usage Questions]
Hi all,
I've got a bit stuck today so reaching out to the community for a bit of help if anybody can assist.
The problem I have is I have 4 tables; 

'Project' this is my top level table that has the project_id, code, name.
'ProjectConstruction' this is a table that holds unique construction information for the Project record (1to1 relationship)
'BuildType' this is a table that contains all the types/categories of building (1tomany, each project might have multiple types)
'BuildIndex' this is my table which i think is effectively my 'secondary' table. it has project_id and build_type_id.

Now if I understand correctly I could build a relationship in Project that would be something like this
`typelist: Mapped[list["BuildType"]] = relationship("BuildType", secondary="optn.build_index")`

But in my case I want the 'typelist' to live/be called from ProjectConstruction and this is where I'm having problems.
I'm adding a simplified table version with just the relevant info below so you can see where i got to.

```
class Project(Base, SystemBase):
    __tablename__ = 'project'
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    code: Mapped[str] = mapped_column(NVARCHAR(40), index=True)
    name: Mapped[str] = mapped_column(NVARCHAR(255), index=True)

class ProjectConstruction(Base, SystemBase):
    __tablename__ = 'construction'
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("proj.project.id"), index=True)
    typelist: Mapped[list["BuildType"]] = relationship("BuildType", secondary="optn.build_index", foreign_keys="[BuildIndex.project_id]")

class BuildType(Base, SystemBase):
    __tablename__ = 'build_type'
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(NVARCHAR(50))

class BuildIndex(Base, SystemBase):
    __tablename__ = 'build_index'
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("proj.project.id"), index=True)
    project: Mapped[Project] = relationship("Project")
    type_id: Mapped[int] = mapped_column(ForeignKey("optn.build_type.id"), index=True)
    type: Mapped["BuildType"] = relationship("BuildType")
```

Any help much appreciated. Thanks!

**Comment by CaselIT** [comment]
Hi,

Keeping that db structure you could use https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html# to map the relationship from Project to ProjectConstruction

**Reply by desktilldawn** [reply]
Thanks for the reply, that's interesting. Not sure it fixes my particular problem though.

I've fixed it this morning by just changing it to a 3 table setup and changing the project_id reference to project_construction_id
So now BuildType is connected to ProjectConstruction using the BuildIndex as an association table.
BuildIndex has a project_construction_id and build_type_id that facilitates the association.
Then Project is connected to ProjectConstruction via the project_id

This introduces the need for the id of ProjectConstruction id in BuildIndex but I think this is ok and at least i have a way forward.

I really like the SQLAlchemy system in general but relationships are really cooking my noodle. I need to find some good tutorials to follow i think as I find the documentation hard reading at times.
